zerver/views/upload.py

```diff
@@ -26,18 +30,14 @@ def json_upload_file(request, user_profile):
     if ((settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024) < user_file._get_size()):
         return json_error(_("File Upload is larger than allowed limit"))
 
-    uri = upload_message_image_through_web_client(request, user_file, user_profile)
+    if is_through_api:
+        uri = upload_message_image_through_api(request, user_file, user_profile)
+    else:
+        uri = upload_message_image_through_web_client(request, user_file, user_profile)
```

Comment:
Hmm, is there a way to avoid having this extra `upload_message_image_through_api` function?  I think it makes the code more confusing.  Would probably be better to just rename the function to e.g. `upload_message_image_from_request` (as its own commit) to make the name clear, and then (as another commit) change the function to do its URL construction using `from django.core.urlresolvers import reverse` (there are a few examples, including one with kwargs, in the views files).  I think if you do that, you shouldn't need this `is_through_api` variable either.